activerecord/lib/active_record/connection_adapters/abstract/query_log_tags.rb

```diff
@@ -0,0 +1,208 @@
+        end
+      end
+
+      # Maintains a user-defined context for all queries and constructs an SQL comment by
+      # calling methods listed in +components+
+      #
+      # Additional information can be added to the context to be referenced by methods
+      # defined in framework or application initializers.
+      #
+      # To add new comment components, define class methods on +QueryLogTagsContext+ in
+      # your application.
+      #
+      #    module ActiveRecord::ConnectionAdapters::QueryLogTags::QueryLogTagsContext
+      #      class << self
+      #        def custom_component
+      #          "custom value"
+      #        end
+      #      end
+      #    end
+      #    ActiveRecord::ConnectionAdapters::QueryLogTags::QueryLogTagsContext.components = []:custom_component]
+      #    ActiveRecord::ConnectionAdapters::QueryLogTags::QueryLogTagsContext.comment
+      #    # /*custom_component:custom value*/
+      #
+      # Default components available for use:
+      #
+      # * +application+
+      # * +pid+
+      # * +socket+
+      # * +db_host+
+      # * +database+
+      # * +line+ (reported via BacktraceCleaner)
+      #
+      # _When included in Rails, ActiveController and ActiveJob components are also defined._
+      #
+      # * +controller+
+      # * +action+
+      # * +job+
+      #
+      # If required due to log truncation, comments can be prepended to the query instead:
+      #
+      #    ActiveRecord::ConnectionAdapters::QueryLogTags.prepend_comment = true
+
+
+      module QueryLogTagsContext
+        mattr_accessor :components, instance_accessor: false, default: [:application]
+        mattr_accessor :cache_query_log_tags, instance_accessor: false, default: true
+        mattr_accessor :backtrace_cleaner, default: ActiveSupport::BacktraceCleaner.new
+        thread_mattr_accessor :cached_comment, instance_accessor: false
+
+        class << self
+          @@backtrace_cleaner.add_silencer { |line| line.match?(/lib\/active_(record|support)/) }
+
+          # Updates the context used to construct the query log tags.
+          # Resets the cached comment if <tt>cache_query_log_tags</tt> is +true+.
+          def update(**options)
+            context.merge!(**options.symbolize_keys)
+            self.cached_comment = nil
+          end
+
+          def tags_available?
+            components.present? || inline_annotations.present?
+          end
+
+          # Returns an SQL comment +String+ containing the query log tags.
+          # Sets and returns a cached comment if <tt>cache_query_log_tags</tt> is +true+.
+          def comment
+            if cache_query_log_tags
+              self.cached_comment ||= uncached_comment
+            else
+              uncached_comment
+            end
+          end
+
+          def uncached_comment
+            content = tag_content
+            if content.present?
+              "/*#{escape_sql_comment(content)}*/"
+            else
+              ""
```

Comment:
What do you think about removing the else and just having the comment be nil if it's not cached rather than empty? Will anything break if we do that?